### ISSUE TITLE
cmd/go: document if the go test and build cache is threadsafe

### ISSUE THREAD

**Issue by alvaroaleman** [golang/go#40461]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.15rc1 linux/amd64
</pre>

### Does this issue reproduce with the latest release?
yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env

</pre></details>

### What did you do?

`go help cache`

### What did you expect to see?

A sentence telling me if the go build and test cache is safe for parallel use by multiple invocation of the `go` binary

**Comment by davecheney** [comment]
I don’t think it is

**Comment by ALTree** [comment]
I know for sure that cleaning the build cache (`go cache -clean`) while building something is not safe. This is discussed in #31948, which also functions as a TODO for possibly making it safe.

**Comment by ALTree** [comment]
cc @bcmills @jayconrod

**Comment by alvaroaleman** [comment]
> I know for sure that cleaning the build cache (go cache -clean) while building something is not safe.

Okay but what about multiple parallel `go build` or `go test` invocations?

**Comment by jayconrod** [comment]
> Okay but what about multiple parallel go build or go test invocations?

This is safe on most file systems. Files are added to the cache atomically, and a global file lock is held when files are evicted. If the file system does not support file locking or atomic renaming, then it may not be safe.

**Comment by bcmills** [comment]
FWIW, the build and test cache does not use file locking, and I don't think it relies on atomic renaming either. (Instead, it writes the expected lengths in metadata files, and overwrites and appends to partial files in-place rather than truncating or deleting them.)

**Comment by bcmills** [comment]
I take that back: we use file-locking in one place in the test cache, to store the expiration date for stale tests.

https://github.com/golang/go/blob/6a167c73977384f3646f4651901fe38347711b10/src/cmd/go/internal/test/test.go#L609-L613

**Comment by networkimprov** [comment]
@alvaroaleman are you hoping to do simultaneous `go build`/`go test` runs of the same project, or different ones, or partially overlapping ones?

**Comment by alvaroaleman** [comment]
@networkimprov the question came up in the context of "We prepopulate and then use one cache per machine that runs many concurrent builds"

**Comment by networkimprov** [comment]
If prepopulated, is your cache read-only?

**Comment by alvaroaleman** [comment]
> If prepopulated, is your cache read-only?

Its not my cache and I believe its not

**Comment by zimmski** [comment]
I do not think that the Go build cache is save for concurrent use or one cannot use it with a CI that uses Kubernetes+Docker. I am maybe hijacking this issue, so please tell me if I should create another one because we are looking for a fix to this which is a major regression for us.

We switched from Go 1.12 to 1.15.6 for one project and are now running into constant non-deterministic build/test problems because of what I think is that Go's cache is not concurrent-safe. Basically our CI does one "go build" over a whole branch and uses one common GOCACHE per CI node. Hence, multiple CI jobs could write to the same cache directory. That part was never a problem even with the newer Go version (i do not know why). After the "go build" CI job multiple jobs run in parallel to do linting and testing (via "go test"). And these run into constant issues e.g. `crypto/tls\n/usr/local/go/src/crypto/tls/common.go:22:2: can't open import: \"net\": open /cache/go-build/87/8713176dd630051cd4b09a7fa534da9a291da82ba5c84c02966f3ee18dc3a206-d: no such file or directory`.

We never had this problem with Go 1.12. The cache just worked. With 1.15.6 it does not. Basically because of this problem we are back to having no cache at all.

We do not run "go clean" at all, so this is not #31948

Any ideas what we should try?

**Comment by bcmills** [comment]
@zimmski, please file a separate issue (including the complete new-issue template). This issue is about documenting the intended behavior — if the behavior does not match what is intended then we need more information in order to diagnose it, and that information is off-topic here.

**Comment by seankhliao** [comment]
Duplicate of #26677
